src/vsr/superblock_free_set_encoded.zig

```diff
@@ -0,0 +1,431 @@
+
+            set.block_addresses[set.block_index] = address;
+            set.block_checksums[set.block_index] = checksum;
+
+            set.grid.?.read_block(
+                .{ .from_local_or_global_storage = open_read_next_callback },
+                &set.read,
+                address,
+                checksum,
+                .{ .cache_read = true, .cache_write = false },
+            );
+        }
+
+        fn open_read_next_callback(read: *Grid.Read, block: Grid.BlockPtrConst) void {
+            const set = @fieldParentPtr(Self, "read", read);
+            assert(set.callback == .open);
+            assert(set.block_index < set.block_count);
+
+            const encoded_words = schema.FreeSetNode.encoded_words(block);
+            const chunk = Chunk.for_block(.{
+                .block_index = set.block_index,
+                .block_count = set.block_count,
+                .free_set_size = set.size,
+            });
+
+            stdx.copy_disjoint(
+                .exact,
+                u8,
+                set.buffer[chunk.start..chunk.end],
+                encoded_words,
+            );
+            set.size_transferred += chunk.size;
+
+            if (schema.FreeSetNode.next(block)) |next| {
+                assert(set.block_index > 0);
+                set.block_index -= 1;
+                set.open_read_next(next.address, next.checksum);
+            } else {
+                assert(set.block_index == 0);
+                set.open_done();
+            }
+        }
+
+        fn open_done(set: *Self) void {
+            assert(set.callback == .open);
+            defer assert(set.callback == .none);
+
+            assert(set.block_index == 0);
+            assert(!set.grid.?.superblock.free_set.opened);
+            assert(set.size_transferred == set.size);
+            assert((set.size > 0) == (set.block_count > 0));
+
+            set.grid.?.superblock.free_set.open(.{
+                .encoded = set.buffer[0..set.size],
+                .block_addresses = set.block_addresses[0..set.block_count],
+            });
+
+            set.grid.?.superblock.free_set.opened = true;
+            assert((set.size > 0) == (set.grid.?.superblock.free_set.count_acquired() > 0));
+
+            const callback = set.callback.open;
+            set.callback = .none;
+            callback(set);
+        }
+
+        pub fn checkpoint(set: *Self, callback: *const fn (set: *Self) void) void {
+            assert(set.callback == .none);
+            defer assert(set.callback == .checkpoint);
+
+            // Checkpoint process is delicate:
+            //   - encode free set,
+            //   - derive the number of blocks required to store the encoding,
+            //   - allocate blocks for the encoding (in the old checkpoint),
+            //   - checkpoint the free set,
+            //   - release the freshly acquired blocks in the new checkpoint.
```

Comment:
To emphasize/check the end state:
```
defer assert(set.grid.?.superblock.free_set.count_released() == set.block_count);
```

Also maybe number the list, since the order is quite important!